src/mozilla_linux_pkg_manager/cli.py

```diff
@@ -168,121 +95,84 @@ def batched(iterable, n):
-    for architecture in architectures:
-        pkg_url = f"{normalized_url}main/binary-{architecture}/Packages"
-        package_data_promises.append(
-            retry_async(
-                fetch_url,
-                args=[pkg_url],
-                attempts=3,
-            )
-        )
-    logging.info("Downloading package data...")
-    package_data_results = await asyncio.gather(*package_data_promises)
-    logging.info("Package data downloaded")
-    logging.info("Unpacking package data")
-    package_data = []
-    for architecture, package_data_result in zip(architectures, package_data_results):
-        logging.info(f"Parsing {architecture} data")
-        parsed_package_data = [
-            yaml.safe_load(raw_package_data)
-            for raw_package_data in package_data_result.split("\n\n")
-        ]
-        package_data.extend(parsed_package_data)
-    logging.info("Package data unpacked")
-    logging.info("Loading nightly package data")
-    firefox_package_data = []
-    for package in package_data:
-        if package and "firefox" in package["Package"]:
-            version, postfix = package["Version"].split("~")
-            package["Gecko-Version"] = GeckoVersion.parse(version)
-            if package["Gecko-Version"].is_nightly:
-                package["Build-ID"] = postfix
-                package["Moz-Build-Date"] = datetime.strptime(
-                    package["Build-ID"], "%Y%m%d%H%M%S"
-                )
-            else:
-                package["Build-Number"] = postfix[len("build") :]
-            firefox_package_data.append(package)
-    nightly_package_data = [
-        package
-        for package in firefox_package_data
-        if package and package["Gecko-Version"].is_nightly
-    ]
-    if not nightly_package_data:
-        logging.info("No nightly_package_data, nothing to do!")
+    packages = await client.list_packages(request=request, retry=ASYNC_RETRY)
+    return packages
+
+
+async def list_versions(package):
+    client = artifactregistry_v1.ArtifactRegistryAsyncClient()
+    request = artifactregistry_v1.ListVersionsRequest(
+        parent=package.name,
+        page_size=1000,
+    )
+    versions = await client.list_versions(request=request, retry=ASYNC_RETRY)
+    return versions
+
+
+async def clean_up(args):
+    logging.info("Pinging repository...")
+    repository = await get_repository(args)
+    logging.info(f"Found repository: {repository.name}")
+    packages = await list_packages(repository)
+    now = datetime.now(UTC)
+    targets = defaultdict(set)
+    pattern = re.compile(args.package)
+    unique_expired_versions = set()
+    all_versions = set()
+
+    start = time.time()
+
+    async for package in packages:
+        name = os.path.basename(package.name)
+        if pattern.match(name):
+            logging.info(f"Looking for expired package versions of {name}...")
+            versions = await list_versions(package)
+            async for version in versions:
```

Comment:
I think you either want to drop the `async` here or use `async for version in list_versions(package)`. As far as I can tell, `versions` is already a simple list as things are written now (because you `await`'ed it). The same thing applies a bit further up with `packages`.

Reply:
[The client returns `async` iterators](https://github.com/googleapis/google-cloud-python/blob/0f194973fb764d15ddb68f77941cc4908bf0a03d/packages/google-cloud-artifact-registry/google/cloud/artifactregistry_v1/services/artifact_registry/pagers.py#L793).
For ex. [ListPackagesAsyncPager](ListPackagesAsyncPager) in the case of `list_versions` (not a plain `list`.)

Reply:
Oh my bad - sorry! Yes, you're absolutely right.